src/sensors/distant.cpp

```diff
@@ -0,0 +1,494 @@
+
+            if (!m_ray_target_shape)
+                Throw("Invalid parameter ray_target, must be a Point3f or a "
+                      "Shape.");
+        } else {
+            Log(Debug, "No target specified.");
+        }
+
+        // Set ray origin
+        if constexpr (OriginType == RayOriginType::Shape) {
+            auto obj           = props.object("ray_origin");
+            m_ray_origin_shape = dynamic_cast<Shape *>(obj.get());
+
+            if (!m_ray_origin_shape)
+                Throw("Invalid parameter ray_origin, must be a Shape.");
+        } else {
+            Log(Debug, "Using bounding sphere for ray origins.");
+        }
+    }
+
+    void set_scene(const Scene *scene) override {
+        m_bsphere = scene->bbox().bounding_sphere();
+        m_bsphere.radius =
+            max(math::RayEpsilon<Float>,
+                m_bsphere.radius * (1.f + math::RayEpsilon<Float>) );
+    }
+
+    template <typename RayVariant>
+    std::pair<RayVariant, Spectrum>
+    sample_ray_impl(Float time, Float wavelength_sample,
+                    const Point2f &film_sample, const Point2f &aperture_sample,
+                    Mask active) const {
+        MTS_MASK_ARGUMENT(active);
+
+        RayVariant ray;
+        ray.time = time;
+
+        // 1. Sample spectrum
+        auto [wavelengths, wav_weight] =
+            sample_wavelength<Float, Spectrum>(wavelength_sample);
+        ray.wavelengths = wavelengths;
+
+        // 2. Set ray direction
+        auto trafo = m_world_transform->eval(time, active);
+        Vector3f v0{ 0.f, 0.f, 1.f };
+
+        if (m_direction_type == RayDirectionType::SampleAll)
+            v0 = warp::square_to_uniform_hemisphere(film_sample);
+
+        else if (m_direction_type == RayDirectionType::SampleWidth) {
+            // Sample directions only in plane generated by X and Z axes
+            auto [s, c] = sincos(math::Pi<ScalarFloat> * film_sample.x());
+            v0.x()      = c;
+            v0.z()      = s;
+        }
+
+        // By default, rays point inwards the target direction,
+        // but the flip_directions parameter allows for switching this
+        // behavior
+        ray.d = m_flip_directions ? trafo.transform_affine(v0)
+                                  : trafo.transform_affine(-v0);
+
+        // 3. Sample ray origin
+        Spectrum ray_weight = 0.f;
+        Point3f ray_target  = m_ray_target_point;
+
+        // 3.1. Sample target point
+        if constexpr (TargetType == RayTargetType::Point) {
+            // Target point selection already handled during init
+            ray_weight = wav_weight;
+        }
+
+        else if constexpr (TargetType == RayTargetType::Shape) {
+            // Use area-based sampling of shape
+            PositionSample3f ps = m_ray_target_shape->sample_position(
+                time, aperture_sample, active);
+            SurfaceInteraction3f si(ps, zero<Wavelength>());
```

Comment:
What is the point of using a `SurfaceInteraction3f` here?

Reply:
No point indeed, I fixed that.